### ISSUE TITLE
Should have recover logic dealing with cached `list.json` on compiler download interruption

### ISSUE THREAD

**Issue by AmagiDDmxh** [nomiclabs/hardhat#1466]
### Step to reproduce
1. run `npx hardhat run scripts/deploy.js --network localhost` which would install new compiler
2. In the middle, interrupt it by press `control` + `c`
3. run `npx hardhat run scripts/deploy.js --network localhost` which would install new compiler again
4. Error prompts
```bash
An unexpected error occurred:

SyntaxError: /Users/.../Library/Caches/hardhat-nodejs/compilers/macosx-amd64/list.json: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at /Users/.../hardhat-hackathon-boilerplate/node_modules/jsonfile/index.js:33:18
    at /Users/.../hardhat-hackathon-boilerplate/node_modules/graceful-fs/graceful-fs.js:123:16
```

During the interruption, write stream on the `list.json` simply turn off instead of recover the previous file, I guess.

**Comment by alcuadrado** [comment]
Thanks for reporting this issue, @AmagiDDmxh! 

I think the problem is that the interruption came in the middle of the `list.json` download. We should download it to a temporal file instead, and then move it, as it's a shorter (if not atomic) operation.

In the meantime, run `npx hardhat clean --global`.

**Comment by kanej** [comment]
I have been able to reproduce the error based on the assumption that the `list.json` file is only partially downloaded:

1. create a new hardhat project based on the basic template
2. Start `npx hardhat node`
3. Run the sample deploy script `npx hardhat run ./scripts/sample-script.js --network localhost` triggering a compile (and compiler download)
4. Edit the `/Users/<user-name>/Library/Caches/hardhat-nodejs/compilers/macosx-amd64/list.json` file to be a partial and invalid json file, so deleting half of it for instance
5. Edit the `contracts/Greeter.sol` to force recompile on next deploy
6. Rerun the sample deploy script `npx hardhat run ./scripts/sample-script.js --network localhost`

That gives the expected error:

```shell
An unexpected error occurred:

SyntaxError: /Users/<user-name>/Library/Caches/hardhat-nodejs/compilers/macosx-amd64/list.json: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at /Users/<user-name>/projects/sandbox/another/node_modules/jsonfile/index.js:33:18
    at /Users/<user-name>/projects/sandbox/another/node_modules/graceful-fs/graceful-fs.js:123:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:71:3)
```

## Approach

Rather than minimising the opportunity for a halt to leave the `list.json` in a bad state, can I suggest we have hardhat recover from a malformed `list.json` by redownloading it and trying to parse it once more. That would cover this case of an interrupted download but guard against malformed `list.json` more generally.

I will push a PR showing the idea.

**Comment by fvictorio** [comment]
Fixed in [`hardhat@2.7.1`](https://github.com/nomiclabs/hardhat/releases/tag/hardhat%402.7.1)
